### PR TITLE
Perform `cfg` attribute processing during macro expansion and fix bugs

### LINKED ISSUE #22250: Process cfg_attr during expansion
```rust
macro_rules! borked {
    () => {
        #[derive(Debug)] pub struct Good;
        #[cfg_attr(not(test), derive(Debug))] pub struct Bad;
    }
}

borked!();

fn main() {
    println!("{:?}", Good);
    println!("{:?}", Bad);
}
```

produces

```
foo.rs:12:22: 12:25 error: the trait `core::fmt::Show` is not implemented for the type `Bad` [E0277]  
foo.rs:12     println!("{:?}", Bad);  
                               ^~~  
```

I think what happens is the following:

* The first cfg stripping pass runs.  At this point the macro body is just a token tree, and there's no way for cfg stripping to work out what `# [ cfg_attr ( not ( test ) ...` means.

* Syntax expansion runs.  The macro emits two items.  One has a `derive` attribute, which matches a `Decorator` in the syntax environment, so that extension also runs.  There's no extension named `cfg_attr` so that attribute has no effect on expansion.

* The post-expansion cfg stripping pass runs.  The second attribute turns into `#[derive(Debug)]` but not in time for syntax expansion to notice it.

Previously, `cfg_attr` *was* a syntax extension, which had numerous drawbacks, but handled this case correctly.

The somewhat obvious fix is to re-run `cfg_attr` processing on the AST subtree produced by every macro expansion, but I'd like to think about this more.  I'd also like to know whether it's something people often run into in practice.

Was [mentioned](https://github.com/rust-lang/rust/issues/19372#issuecomment-74131263) on #19372.  cc @sfackler, @SergioBenitez

### LINKED ISSUE #31369: cfg attributes appear are improperly applied inside of `include!`
Example to reproduce:

```rust
// src/lib.rs
include!("foo.rs");

// foo.rs
#[cfg(feature = "never")]
include!("error.rs");

// error.rs
omg i'm a syntax error
```

This should successfully compile, but instead will have a syntax error. It should behave identically to what would occur if you inlined the first `include!`, (which does compile successfully)

```rust
// src/lib.rs
#[cfg(feature = "never")]
include!("error.rs");

// error.rs
omg i'm a syntax error
```

### PR DESCRIPTION
**Pull request by jseyfried**
This PR refactors `cfg` attribute processing and fixes bugs. More specifically:
- It merges gated feature checking for stmt/expr attributes, `cfg_attr` processing, and `cfg` processing into a single fold.
  - This allows feature gated `cfg` variables to be used in `cfg_attr` on unconfigured items. All other feature gated attributes can already be used on unconfigured items.
- It performs `cfg` attribute processing during macro expansion instead of after expansion so that macro-expanded items are configured the same as ordinary items. In particular, to match their non-expanded counterparts,
  - macro-expanded unconfigured macro invocations are no longer expanded,
  - macro-expanded unconfigured macro definitions are no longer usable, and
  - feature gated `cfg` variables on macro-expanded macro definitions/invocations are now errors.

This is a [breaking-change]. For example, the following would break:

``` rust
macro_rules! m {
    () => {
        #[cfg(attr)]
        macro_rules! foo { () => {} }
        foo!(); // This will be an error

        macro_rules! bar { () => { fn f() {} } }
        #[cfg(attr)] bar!(); // This will no longer be expanded ...
        fn g() { f(); } // ... so that `f` will be unresolved.

        #[cfg(target_thread_local)] // This will be a gated feature error
        macro_rules! baz { () => {} }
    }
}

m!();
```

**EDIT:** fixes #22250, fixes #31369.

r? @nrc

### CODE REVIEW

File: src/libsyntax/ext/expand.rs [hunk 1]
```diff
@@ -69,16 +68,12 @@ pub fn expand_expr(e: P<ast::Expr>, fld: &mut MacroExpander) -> P<ast::Expr> {
             };
 
             // Keep going, outside-in.
-            let fully_expanded = fld.fold_expr(expanded_expr);
-            let span = fld.new_span(span);
+            let fully_expanded = fld.fold_expr(expanded_expr).map(|e| {
+                ast::Expr { span: Span { expn_id: fld.cx.backtrace(), ..span }, ..e }
```
Comment (jseyfried) [inline src/libsyntax/ext/expand.rs]: cf #33712

**Comment by jseyfried** [comment]
cc @nikomatsakis

**Comment by jseyfried** [comment]
Rebased.

File: src/libsyntax/config.rs [hunk 2]
```diff
@@ -19,13 +19,31 @@ use ptr::P;
 
 use util::small_vector::SmallVector;
 
+pub trait CfgFolder: fold::Folder {
+    fn in_cfg(&mut self, attrs: &[ast::Attribute]) -> bool;
+    fn visit_unconfigurable_expr(&mut self, _expr: &ast::Expr) {}
```
Comment (nrc) [inline src/libsyntax/config.rs]: Could you document these methods please?

**Comment by nrc** [comment]
r+ with the comment added

**Comment by jseyfried** [comment]
@nrc I made sure we process `cfg_attr` attributes on non-optional expressions and added comments.

**Comment by jseyfried** [comment]
@nrc also, some process questions:
 - Should this PR be tagged `relnotes`? (c.f. [#33458](https://github.com/rust-lang/rust/issues/33458#issuecomment-219183152))
 - Do you intend "r+ with the comment added" to have the same meaning as "r=me with the comment added"? Currently, I don't approve on behalf of a reviewer unless they say "r=me".

**Comment by jseyfried** [comment]
Rebased.

**Comment by nrc** [comment]
@bors: r+

@jseyfried yes and yes

**Comment by jseyfried** [comment]
Fixes #22250.

**Comment by jseyfried** [comment]
Fixes #31369.

**Pull request merged**